Add organization on project page

subinasr opened the pull request:
## Stakeholders Modal on Project page

- Drag and drop organization 
- Add / Remove organization 
- Create new organization

fixes https://github.com/the-deep/client/issues/1095

tnagorra committed:
Add organization on project page

AdityaKhatri commented on src/components/input/StakeholderModal/OrganizationList/index.js:
```diff
@@ -180,6 +180,7 @@ export default class OrganizationList extends React.PureComponent {
             isCapped,
             organizations,
         } = this.filterOrganization(sources.organizations, searchValue);
+        console.log(OrganizationList);
```
remove this

AdityaKhatri commented on src/components/input/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,200 @@
+        query: { fields: ['id', 'title'] },
+        method: methods.GET,
+        onMount: true,
+        onSuccess: ({ response }) => {},
```
onSuccess is not required here

AdityaKhatri commented on src/components/input/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,200 @@
+});
+
+const propTypes = {
+    closeModal: PropTypes.func.isRequired,
```
onChange, requests and value is missing in propTypes

AdityaKhatri commented on src/components/input/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,200 @@
+                    value={faramValues}
+                    error={faramErrors}
+                >
+                    <ModalHeader title="Stakeholders" />
```
Add `close` button on header on the right that closes the modal

AdityaKhatri replied:
use translation

AdityaKhatri commented on src/components/input/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,200 @@
+                    <ModalBody className={styles.modalBody}>
+                        <MultiSelectInput
+                            faramElementName="leadOrganization"
+                            label="Lead Organization"
```
use translation

AdityaKhatri commented on src/components/input/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,200 @@
+                        />
+                        <MultiSelectInput
+                            faramElementName="internationalPartners"
+                            label="International Partners"
```
use translation

AdityaKhatri commented on src/components/input/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,200 @@
+                        />
+                        <MultiSelectInput
+                            faramElementName="donors"
+                            label="Donors"
```
use translation

AdityaKhatri commented on src/components/input/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,200 @@
+                        />
+                        <MultiSelectInput
+                            faramElementName="nationalPartners"
+                            label="National Partners"
```
use translation

AdityaKhatri commented on src/components/input/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,200 @@
+                        />
+                        <MultiSelectInput
+                            faramElementName="government"
+                            label="Government"
```
use translation

AdityaKhatri commented on src/components/input/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,200 @@
+                        <DangerButton
+                            onClick={closeModal}
+                        >
+                            Cancel
```
use translation

AdityaKhatri commented on src/components/input/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,200 @@
+                        <SuccessButton
+                            type="submit"
+                        >
+                            Save
```
use translation

AdityaKhatri commented on src/views/Project/Details/General/index.js:
```diff
@@ -118,6 +135,8 @@ export default class ProjectDetailsGeneral extends PureComponent {
             projectId,
         });
+
+        console.warn('faramValues', faramValues);
```
remove this

AdityaKhatri commented on src/views/Project/Details/General/index.js:
```diff
@@ -266,6 +294,16 @@ export default class ProjectDetailsGeneral extends PureComponent {
+                                    />
+                                }
+                            >
+                                Stakeholder
```
use translation

AdityaKhatri commented on src/views/Project/Details/General/index.js:
```diff
@@ -31,6 +35,8 @@ import {
     routeUrlSelector,
 } from '#redux';
 
+import { getProps } from '#entities/editAry';
```
this is not required

AdityaKhatri commented on src/views/Project/Details/General/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,217 @@
+};
+
+const defaultProps = {
+    closeModal: () => {},
```
we should keep this as undefined and check during every usuage

AdityaKhatri commented on src/views/Project/Details/General/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,217 @@
+
+const defaultProps = {
+    closeModal: () => {},
+    onChange: () => {},
```
we should keep this as undefined and check during every usuage

AdityaKhatri commented on src/views/Project/Details/General/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,217 @@
+    closeModal: () => {},
+    onChange: () => {},
+    value: {},
+    requests: {},
```
requests is a required field

AdityaKhatri commented on src/views/Project/Details/General/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,217 @@
+const defaultProps = {
+    closeModal: () => {},
+    onChange: () => {},
+    value: {},
```
value is also a required field

samshara committed:
Add organization to project

Implement add organizations to a project. List all organizations belonging
to a project.Add search functionality on organizations and drag and drop
organization to each organization type for a project. Create a new organization
if not present.

frozenhelium committed:
Fix request for project stakeholders

frozenhelium commented:
@subinasr Please closely review and inspect this PR. It'll be really useful.

tnagorra commented on src/views/Project/Details/General/StakeholdersModal/OrganizationField/index.tsx:
```diff
@@ -0,0 +1,130 @@
+
+interface Props {
+    containerClassName?: string;
+    onChange: (values: (number | undefined)[]) => void;
```
The typing should be 
`values: number[] | undefined`

tnagorra commented on src/views/Project/Details/General/StakeholdersModal/OrganizationField/index.tsx:
```diff
@@ -0,0 +1,130 @@
+        }
+    }, [dragEnterCount, setDragEnterCount, setIsBeingDraggedOver]);
+
+    const handleDragOver = (e: React.DragEvent<HTMLDivElement>) => {
+        e.preventDefault();
+    };
+    const handleDrop = useCallback((e: React.DragEvent<HTMLDivElement>) => {
+        e.preventDefault();
+        const data = e.dataTransfer.getData('text');
+        try {
+            const parsedData = JSON.parse(data);
+            if (parsedData && parsedData.organizationId) {
+                const { organizationId } = parsedData;
+
+                if (!value) {
+                    onChange([organizationId]);
+                } else if (value.findIndex(v => v === organizationId) === -1) {
+                    const intercept = false;
+
+                    if (intercept) {
+                        setShowConfirmation(true);
+                        setDroppedOrganizationId(organizationId);
+                    } else {
+                        onChange([...value, organizationId]);
+                    }
+                }
+            }
+        } catch (ex) {
+            console.warn('Only organizations supported');
+        }
+        setDragEnterCount(0);
+        setIsBeingDraggedOver(false);
+    }, [
+        value,
+        onChange,
+        setDragEnterCount,
+        setIsBeingDraggedOver,
+        setShowConfirmation,
+        setDroppedOrganizationId,
+    ]);
```
Create a common hook to handle drag logic

tnagorra commented on src/views/Project/Details/General/StakeholdersModal/SearchOrganization/index.tsx:
```diff
@@ -0,0 +1,127 @@
+
+    const [searchText, setSearchValue] = useState<string|undefined>(undefined);
+
+    useEffect(() => {
+        const timeoutId = setTimeout(() => {}, 1000);
+        return () => clearTimeout(timeoutId);
+    }, [searchText]);
```
What does this do?

frozenhelium replied:
@samshara

tnagorra commented on src/views/Project/Details/General/StakeholdersModal/SearchOrganization/index.tsx:
```diff
@@ -0,0 +1,127 @@
+        return () => clearTimeout(timeoutId);
+    }, [searchText]);
+
+    const handleChange = useCallback((value?: string) => {
```
```suggestion
    const handleChange = useCallback((value: string | undefined) => {
```

tnagorra commented on src/views/Project/Details/General/StakeholdersModal/SearchOrganization/index.tsx:
```diff
@@ -0,0 +1,127 @@
+        }
+    }, [setSearchValue, handleSearch]);
+
+    const renderEmpty = () => (
+        <div className={styles.emptyComponent}>
+            { organizationList.length === 0 && !searchText ? (
+                _ts('assessment.metadata.stakeholder', 'typeToSearchOrganizationMessage')
+            ) : (
+                _ts('assessment.metadata.stakeholder', 'noResultsFoundMessage')
+            )}
+        </div>
+    );
```
This should be outside the component.
We should have two Message components and switch between them

frozenhelium committed:
Improve project stakeholder layout in project page

samshara committed:
Refactor drag and drop logic into reusable hook

tnagorra commented on src/hooks/useDropHandler.tsx:
```diff
@@ -0,0 +1,52 @@
+    const [dropping, setDropping] = React.useState(false);
+    const dragEnterCount = React.useRef(0);
+
+    const onDragOver = (e: React.DragEvent<HTMLDivElement>) => {
+        e.preventDefault();
+    };
+
+    const onDragEnter = (e: React.DragEvent<HTMLDivElement>) => {
+        if (dragEnterCount.current === 0) {
+            setDropping(true);
+
+            dragStartHandler(e);
+        }
+        dragEnterCount.current += 1;
+    };
+    const onDragLeave = () => {
+        dragEnterCount.current -= 1;
+        if (dragEnterCount.current === 0) {
+            setDropping(false);
+        }
+    };
+
+    const onDrop = (e: React.DragEvent<HTMLDivElement>) => {
+        dragEnterCount.current = 0;
+        setDropping(false);
+
```
We should use useCallback on these functions.

tnagorra commented on src/views/Project/Details/General/StakeholdersModal/OrganizationField/index.tsx:
```diff
@@ -0,0 +1,119 @@
+    ]);
+
+    const handleConfirmation = useCallback((confirm: boolean) => {
+        console.warn('handleConfirmation called');
```
Let's not leave console.warn

tnagorra commented on src/views/Project/Details/General/StakeholdersModal/SearchOrganization/index.tsx:
```diff
@@ -0,0 +1,145 @@
+}
+
+const organizationKeySelector = (d: Organization) => d.id;
+// eslint-disable-next-line
```
Oh no no no no!

tnagorra commented on src/views/Project/Details/General/index.js:
```diff
@@ -79,6 +92,91 @@ const mapDispatchToProps = dispatch => ({
+        return null;
+    }
+
+    const organizationType = fields.find(v => v.faramElementName === data[0].organizationType);
```
I think we should memoize this!

tnagorra commented on src/views/Project/Details/General/index.js:
```diff
@@ -150,6 +249,19 @@ export default class ProjectDetailsGeneral extends PureComponent {
+            o => o,
+        );
+
+        const value = values[key] || [];
```
We should not pass emptylist that is always created

frozenhelium committed:
Fix an issue caused by non uniqueness of project organization

AdityaKhatri commented on src/views/Project/Details/General/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,308 @@
+            logo: o.logoUrl,
+        }));
+
+        const uniqueOptions = [...organizationOptions];
+        uniqueAppend(uniqueOptions, organizationOptions, o => o.id);
+        /*
+        const uniqueOptions = unique(
+            [
+                ...transformedOptions,
+                ...organizationOptions,
+
+            ].filter(isDefined),
+            option => option.id,
+        );
+        */
```
@frozenhelium , should we not include transformedOptions in unique options?
What was the issue with commented out code?
Also, let's avoid writing functions that mutate the values. Was the mutation necessary?

tnagorra commented on src/views/Project/Details/General/StakeholdersModal/OrganizationField/index.tsx:
```diff
@@ -0,0 +1,118 @@
+        try {
+            const data = e.dataTransfer.getData('text');
+            const parsedData = JSON.parse(data);
+            if (parsedData && parsedData.organizationId) {
+                const { organizationId } = parsedData;
+
+                if (!value) {
+                    onChange([organizationId]);
+                } else if (value.findIndex(v => v === organizationId) === -1) {
+                    const intercept = false;
+
+                    if (intercept) {
+                        setShowConfirmation(true);
+                        setDroppedOrganizationId(organizationId);
+                    } else {
+                        onChange([...value, organizationId]);
+                    }
+                }
+            }
```
We should break early instead of increasing nesting.

tnagorra commented on src/views/Project/Details/General/StakeholdersModal/OrganizationField/index.tsx:
```diff
@@ -0,0 +1,118 @@
+    ]);
+
+    const handleConfirmation = useCallback((confirm: boolean) => {
+        if (confirm) {
+            if (droppedOrganizationId) {
+                onChange([
+                    ...value,
+                    droppedOrganizationId,
+                ]);
+            }
+        }
```
better to merge both if into one

tnagorra commented on src/views/Project/Details/General/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,307 @@
+    let organizations = [];
+    Object.entries(faramValues).forEach(([key, values]) => {
+        const organizationList = values
+            .map(v => options.find(o => o.id === v))
```
It's better if accessed through a map

tnagorra commented on src/views/Project/Details/General/StakeholdersModal/index.js:
```diff
@@ -0,0 +1,307 @@
+    return organizations;
+};
+
+function uniqueAppend(list, elements, keySelector) {
+    const map = listToMap(list, keySelector);
+    elements.forEach((el) => {
+        if (!isDefined(map[keySelector(el)])) {
+            list.push(el);
+        }
+    });
+}
```
We already have a unique function.why not just use unique([...oldArray, ...newArray], keySelector])

frozenhelium committed:
Remove uniqueAppend and use unique from fujs

samshara committed:
Fix coding patterns

tnagorra commented on src/views/Project/Details/General/StakeholdersModal/OrganizationField/index.tsx:
```diff
@@ -0,0 +1,118 @@
+            if (!value) {
+                onChange([organizationId]);
+            } else if (value.findIndex(v => v === organizationId) === -1) {
+                const intercept = false;
+
+                if (intercept) {
+                    setShowConfirmation(true);
+                    setDroppedOrganizationId(organizationId);
+                } else {
+                    onChange([...value, organizationId]);
+                }
```
The `if` branch is not reachable (i think we should remove this)

samshara committed:
Remove Confirm
